File: source/ch/cyberduck/ui/cocoa/ErrorController.java
package ch.cyberduck.ui.cocoa;

/*
 *  Copyright (c) 2008 David Kocher. All rights reserved.
 *  http://cyberduck.ch/
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  Bug fixes, suggestions and comments should be sent to:
 *  dkocher@cyberduck.ch
 */

import ch.cyberduck.core.threading.BackgroundException;
import ch.cyberduck.ui.cocoa.application.NSTextField;
import ch.cyberduck.ui.cocoa.application.NSView;
import ch.cyberduck.ui.cocoa.foundation.NSAttributedString;

import org.apache.log4j.Logger;

/**
 * @version $Id:$
 */
public class ErrorController extends BundleController {
    private static Logger log = Logger.getLogger(TaskController.class);

    @Outlet
    private NSTextField hostField;

    public void setHostField(NSTextField hostField) {
        this.hostField = hostField;
        if(null == failure.getPath()) {
            this.hostField.setAttributedStringValue(
                    NSAttributedString.attributedStringWithAttributes(failure.getSession().getHost().toURL(), FIXED_WITH_FONT_ATTRIBUTES));
        }
        else {
            this.hostField.setAttributedStringValue(
                    NSAttributedString.attributedStringWithAttributes(failure.getPath().getAbsolute(), FIXED_WITH_FONT_ATTRIBUTES));
        }
    }

    @Outlet
    private NSTextField descriptionField;

    public void setDescriptionField(NSTextField descriptionField) {
        this.descriptionField = descriptionField;
        this.descriptionField.setSelectable(true);
        this.descriptionField.setAttributedStringValue(
                NSAttributedString.attributedStringWithAttributes(failure.getDetailedCauseMessage(), TRUNCATE_MIDDLE_ATTRIBUTES));
    }

    @Outlet
    private NSTextField errorField;

    public void setErrorField(NSTextField errorField) {
        this.errorField = errorField;
        this.errorField.setSelectable(true);
        this.errorField.setAttributedStringValue(
                NSAttributedString.attributedStringWithAttributes(failure.getReadableTitle() + ": " + failure.getMessage(), TRUNCATE_MIDDLE_ATTRIBUTES));
    }

    @Outlet
    private NSView view;

    public void setView(NSView view) {
        this.view = view;
    }

    @Override
    public NSView view() {
        return view;
    }

    private BackgroundException failure;

    public ErrorController(BackgroundException e) {
        this.failure = e;
        this.loadBundle();
    }

    /**
     * @return
     */
    public String getTooltip() {
        return failure.getReadableTitle();
    }

    @Override
    protected String getBundleName() {
        return "Error";
    }
}

File: source/ch/cyberduck/ui/controller/ErrorController.cs
// 
// Copyright (c) 2010 Yves Langisch. All rights reserved.
// http://cyberduck.ch/
// 
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// Bug fixes, suggestions and comments should be sent to:
// yves@cyberduck.ch
// 
using System.Collections.Generic;
using ch.cyberduck.core;
using ch.cyberduck.core.threading;

namespace Ch.Cyberduck.Ui.Controller
{
    internal class ErrorController : WindowController<IErrorView>
    {
        public ErrorController(IErrorView view, IEnumerable<BackgroundException> exceptions, string transcript)
        {
            View = view;

            View.SetModel(exceptions);
            View.ModelErrorMessageGetter =
                delegate(object rowObject)
                    {
                        BackgroundException failure = (BackgroundException) rowObject;
                        return GetReadableTitle(failure) + ": " + failure.getMessage();
                    };

            View.ModelHostGetter = delegate(object rowObject)
                                       {
                                           BackgroundException failure = (BackgroundException) rowObject;
                                           if (null == failure.getPath())
                                           {
                                               return failure.getSession().getHost().toURL();
                                           }
                                           return failure.getPath().getAbsolute();
                                       };
            View.ModelDescriptionGetter =
                delegate(object rowObject) { return GetDetailedCauseMessage((BackgroundException) rowObject); };

            bool log = !string.IsNullOrEmpty(transcript);
            View.Transcript = transcript;
            View.TranscriptVisible = log &&
                                     Preferences.instance().getBoolean("alert.toggle.transcript");
            View.TranscriptEnabled = log;
            View.ToggleTranscriptEvent += View_ToggleTranscriptEvent;
        }

        private void View_ToggleTranscriptEvent()
        {
            View.TranscriptVisible = !View.TranscriptVisible;
            Preferences.instance().setProperty("alert.toggle.transcript", View.TranscriptVisible);
        }

        private string GetReadableTitle(BackgroundException e)
        {
            return e.getReadableTitle();
        }

        private string GetDetailedCauseMessage(BackgroundException e)
        {
            return e.getDetailedCauseMessage();
        }
    }
}